src/testdrive/src/action/mysql/connect.rs

```diff
@@ -27,10 +29,18 @@ pub async fn run_connect(
         .map_err(|_| anyhow!("Unable to parse MySQL URL {}", url))?;
     let opts = mysql_async::OptsBuilder::from_opts(opts_url).pass(password.clone());
     let pool = mysql_async::Pool::new(opts);
-    let conn = pool
-        .get_conn()
-        .await
-        .map_err(|_| anyhow!("Unable to connect to MySQL server at {}", url))?;
+
+    let conn = Retry::default()
+        .max_duration(state.default_timeout)
+        .retry_async_canceling({
+            |_| async {
+                let url2 = url.clone();
+                pool.get_conn().await.map_err(|err| {
+                    anyhow!("Unable to connect to MySQL server at {}: {:?}", url2, err)
+                })
+            }
+        })
+        .await?;
```

Comment:
Agreed, I spent all day yesterday struggling to replicate the error on this -- the [healthcheck](https://github.com/MaterializeInc/materialize/blob/main/misc/python/materialize/mzcompose/services/mysql.py#L41) appears to work fine and is the [recommended](https://stackoverflow.com/questions/42567475/docker-compose-check-if-mysql-connection-is-ready) way to validate mysql is ready, but for some reason the first testdrive `mysql-connect` call would hit a 'Connection Refused' error on the first attempt when run on buildkite but not locally.

Then I noticed that the `postgres-connect` testdrive client [implemented](https://github.com/MaterializeInc/materialize/blob/main/src/testdrive/src/util/postgres.rs#L56) this Retry behavior so I copied over the same implementation to make this work.

However I've now since removed that `mysql-connect` call from the test-case since we removed the `database` option. I could remove this retry behavior but I suspect that the next test-case we add will hit the same thing. I have a feeling something is different about the docker environment in buildkite vs my local one

Reply:
Actually I just realized I'm hitting this again now that I removed the `mysql-connect` call and the first connection to MySQL is the Create Connection validation https://buildkite.com/materialize/tests/builds/71613#018c8435-29e8-456f-8d67-a360c620d38f which doesn't have this retry 🤦🏽 . Will keep digging

Reply:
Figured it out! I updated the healthcheck to use a TCP connection for validation -- it was defaulting to using a unix-socket to check if MySQL was ready to accept connections which must happen quicker than when using TCP.

Reply:
I think it may still make sense to leave this retry logic here to match the postgres behavior, wdyt?

Reply:
~I think the healthcheck is broken! According to this SO post `mysqladmin --ping` can return a false positive if the server is running but not accepting connections: https://stackoverflow.com/questions/42567475/docker-compose-check-if-mysql-connection-is-ready#comment75589721_42757250~

~The timestamps seem to bear that out. In services.log...~

```
mysql-cdc-mysql-1         | 2023-12-19T22:52:53.631782Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Bind-address: '::' port: 33060, socket: /var/run/mysqld/mysqlx.sock
mysql-cdc-mysql-1         | 2023-12-19T22:52:53.631857Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.35'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.
```

~In testdrive...~

```
> CREATE SECRET [query truncated on purpose so as to not reveal the secret in the log]
rows match; continuing at ts 1703026372.2337284
> CREATE CONNECTION mysq TO MYSQL (
  HOST mysql,
  USER root,
  PASSWORD SECRET mysqlpass
)
query error matches; continuing
Connecting to PostgreSQL server at postgres://mz_system:materialize@materialized:6877...
>> ALTER SYSTEM SET enable_mysql_source = true
> CREATE CONNECTION mysq TO MYSQL (
  HOST mysql,
  USER root,
  PASSWORD SECRET mysqlpass
)
```

~1703026372.2337284 is 22:52:52.233, so testdrive is trying to connect (or, at least, is a few commands away from trying to connect) 1.5 seconds before MySQL is accepting connections.~

Scratch all the above; I wrote it up before you mentioned you'd tracked it down. `--protocol=TCP` is a great, great catch—nice work! 🙇🏽 

> I think it may still make sense to leave this retry logic here to match the postgres behavior, wdyt?

I'd leave it without the retry. @def- recently added the retry loop in https://github.com/MaterializeInc/materialize/pull/22300/files#r1354183675 and I'm not entirely sure why! @def-, is it possible that you could use `tokio::time::timeout` in the PostgreSQL connection logic rather than `retry_async_canceling`?